File: lib/data/models/note.dart
import 'package:equatable/equatable.dart';

class Note extends Equatable {
  final String id;
  final String userId;
  final String? notebookId;
  final String title;
  final String content;
  final String contentType; // plain, rich, markdown
  final bool isPinned;
  final bool isArchived;
  final bool isDeleted;
  final DateTime createdAt;
  final DateTime updatedAt;
  final List<String> tagNames;
  final List<NoteVersion> versions;

  const Note({
    required this.id,
    required this.userId,
    this.notebookId,
    required this.title,
    required this.content,
    this.contentType = 'plain',
    this.isPinned = false,
    this.isArchived = false,
    this.isDeleted = false,
    required this.createdAt,
    required this.updatedAt,
    this.tagNames = const [],
    this.versions = const [],
  });

  factory Note.fromJson(Map<String, dynamic> json) {
    return Note(
      id: json['id'] as String,
      userId: json['user_id'] as String,
      notebookId: json['notebook_id'] as String?,
      title: json['title'] as String? ?? 'Untitled',
      content: json['content'] as String? ?? '',
      contentType: json['content_type'] as String? ?? 'plain',
      isPinned: json['is_pinned'] as bool? ?? false,
      isArchived: json['is_archived'] as bool? ?? false,
      isDeleted: json['is_deleted'] as bool? ?? false,
      createdAt: DateTime.parse(json['created_at'] as String),
      // Prefer the server's LWW basis (edited_at = when a human last edited)
      // over updated_at (= when the server applied it). Locally, updatedAt
      // always means "last edit time", so merges compare edit-vs-edit and an
      // older edit that merely synced later can't clobber a newer local edit.
      updatedAt: DateTime.tryParse(json['edited_at'] as String? ?? '') ??
          DateTime.parse(json['updated_at'] as String),
      tagNames:
          (json['tags'] as List<dynamic>?)
              ?.map(
                (t) => (t is Map<String, dynamic>)
                    ? (t['name'] as String? ?? '')
                    : t.toString(),
              )
              .toList() ??
          [],
      versions:
          (json['versions'] as List<dynamic>?)
              ?.map((v) => NoteVersion.fromJson(v as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'user_id': userId,
    'notebook_id': notebookId,
    'title': title,
    'content': content,
    'content_type': contentType,
    'is_pinned': isPinned,
    'is_archived': isArchived,
    'is_deleted': isDeleted,
    'created_at': createdAt.toIso8601String(),
    'updated_at': updatedAt.toIso8601String(),
    'tags': tagNames,
  };

  /// Sentinel distinguishing "not passed" from an explicit null in [copyWith],
  /// so a note can be moved OUT of a notebook (notebookId: null).
  static const Object _unset = Object();

  Note copyWith({
    String? title,
    String? content,
    String? contentType,
    Object? notebookId = _unset,
    bool? isPinned,
    bool? isArchived,
    bool? isDeleted,
    DateTime? updatedAt,
    List<String>? tagNames,
    List<NoteVersion>? versions,
  }) {
    return Note(
      id: id,
      userId: userId,
      notebookId: identical(notebookId, _unset)
          ? this.notebookId
          : notebookId as String?,
      title: title ?? this.title,
      content: content ?? this.content,
      contentType: contentType ?? this.contentType,
      isPinned: isPinned ?? this.isPinned,
      isArchived: isArchived ?? this.isArchived,
      isDeleted: isDeleted ?? this.isDeleted,
      createdAt: createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      tagNames: tagNames ?? this.tagNames,
      versions: versions ?? this.versions,
    );
  }

  @override
  List<Object?> get props => [
    id,
    userId,
    notebookId,
    title,
    content,
    contentType,
    isPinned,
    isArchived,
    isDeleted,
    createdAt,
    updatedAt,
    tagNames,
  ];
}

class NoteVersion extends Equatable {
  final String id;
  final String title;
  final String content;
  final String contentType;
  final int versionNumber;
  final DateTime createdAt;

  const NoteVersion({
    required this.id,
    required this.title,
    required this.content,
    required this.contentType,
    required this.versionNumber,
    required this.createdAt,
  });

  factory NoteVersion.fromJson(Map<String, dynamic> json) {
    return NoteVersion(
      id: json['id'] as String,
      title: json['title'] as String? ?? '',
      content: json['content'] as String? ?? '',
      contentType: json['content_type'] as String? ?? 'plain',
      versionNumber: json['version_number'] as int? ?? 1,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  @override
  List<Object?> get props => [id, versionNumber, createdAt];
}

File: backend/app/models/note.py
import uuid
from datetime import datetime
from enum import Enum as PyEnum
from sqlalchemy import String, Integer, Boolean, DateTime, ForeignKey, Text, Enum, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.database import Base


class ContentType(str, PyEnum):
    PLAIN = "plain"
    RICH = "rich"
    MARKDOWN = "markdown"


class Note(Base):
    __tablename__ = "notes"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    notebook_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("notebooks.id", ondelete="SET NULL"), nullable=True, index=True)
    title: Mapped[str] = mapped_column(String(500), default="Untitled", nullable=False)
    content: Mapped[str] = mapped_column(Text, default="", nullable=False)
    content_type: Mapped[ContentType] = mapped_column(Enum(ContentType), default=ContentType.PLAIN, nullable=False)
    is_pinned: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    is_archived: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False, index=True)
    is_deleted: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False, index=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    # The client's own last-edit time, used ONLY for sync last-write-wins
    # comparison. Kept separate from updated_at (which is the server-controlled
    # sync cursor) so conflict resolution compares edit-time-vs-edit-time and a
    # note synced later can't wrongly lose to one that was edited earlier.
    edited_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    deleted_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    # Relationships
    user: Mapped["User"] = relationship("User", back_populates="notes")
    notebook: Mapped["Notebook | None"] = relationship("Notebook", back_populates="notes")
    # versions/files are deliberately NOT eager-loaded. They used to be
    # lazy="selectin", which auto-loaded the FULL version history (up to
    # _MAX_VERSIONS snapshots, each carrying the entire note body) and every
    # attachment on EVERY note query — including the note-list screen and every
    # sync pull, neither of which uses them. That bloated those responses badly.
    #   versions -> "noload": accessing it returns [] (so the list/get response
    #     still carries a "versions" key, unchanged on the wire) UNLESS a query
    #     opts in via selectinload(Note.versions) — get_note does, for history.
    #   files -> "raise": nothing serializes note.files except the .oblix export,
    #     which loads it explicitly; raising turns any accidental access into a
    #     loud error instead of a silent per-note extra query.
    versions: Mapped[list["NoteVersion"]] = relationship("NoteVersion", back_populates="note", lazy="noload", cascade="all, delete-orphan", order_by="NoteVersion.version_number")
    tags: Mapped[list["NoteTag"]] = relationship("NoteTag", back_populates="note", lazy="selectin", cascade="all, delete-orphan")
    files: Mapped[list["File"]] = relationship("File", back_populates="note", lazy="raise")

    def __repr__(self) -> str:
        return f"<Note {self.title}>"


class NoteVersion(Base):
    __tablename__ = "note_versions"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    note_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("notes.id", ondelete="CASCADE"), nullable=False, index=True)
    title: Mapped[str] = mapped_column(String(500), nullable=False)
    content: Mapped[str] = mapped_column(Text, default="", nullable=False)
    content_type: Mapped[ContentType] = mapped_column(Enum(ContentType), nullable=False)
    version_number: Mapped[int] = mapped_column(Integer, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    # Relationships
    note: Mapped["Note"] = relationship("Note", back_populates="versions")

    def __repr__(self) -> str:
        return f"<NoteVersion note={self.note_id} v{self.version_number}>"